File: lib/src/main/java/com/nytimes/android/external/registerlib/ServiceIntentHelper.java
package com.nytimes.android.external.registerlib;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;

import java.util.List;

/**
 * Utility class added to address API 21 issue with calling service with implicit intent
 *
 */
public final class ServiceIntentHelper {

    private ServiceIntentHelper() {
        // utility class
    }

    /***
     * Android L (lollipop, API 21) introduced a new problem when trying to invoke implicit intent,
     * "java.lang.IllegalArgumentException: Service Intent must be explicit"
     *
     * If you are using an implicit intent, and know only 1 target would answer this intent,
     * This method will help you turn the implicit intent into the explicit form.
     *
     * Inspired from SO answer: http://stackoverflow.com/a/26318757/1446466
     * @param context
     * @param implicitIntent - The original implicit intent
     * @return Explicit Intent created from the implicit original intent
     */
    public static Intent createExplicitFromImplicitIntent(Context context, Intent implicitIntent) {
        ResolveInfo resolveInfo = getResolveInfo(context, implicitIntent);
        if (resolveInfo == null) {
            return null;
        }

        ComponentName component =  getComponentFromServiceInfo(resolveInfo);

        return makeExplicitIntent(implicitIntent, component);
    }

    // Retrieve service that can match the given intent
    private static ResolveInfo getResolveInfo(Context context, Intent implicitIntent) {
        PackageManager packageManager = context.getPackageManager();
        List<ResolveInfo> resolveInfoList = packageManager.queryIntentServices(implicitIntent, 0);

        // Make sure only one match was found
        if (resolveInfoList == null || resolveInfoList.size() != 1) {
            return null;
        }
        return resolveInfoList.get(0);
    }

    // Get component info and create ComponentName
    private static ComponentName getComponentFromServiceInfo(ResolveInfo serviceInfo) {
        String packageName = serviceInfo.serviceInfo.packageName;
        String className = serviceInfo.serviceInfo.name;
        return new ComponentName(packageName, className);
    }

    private static Intent makeExplicitIntent(Intent implicitIntent, ComponentName component) {
        // Create a new intent. Use the old one for extras and such reuse
        Intent explicitIntent = new Intent(implicitIntent);

        // Set the component to be explicit
        explicitIntent.setComponent(component);

        return explicitIntent;
    }
}

File: lib/src/main/java/com/nytimes/android/external/registerlib/ServiceIntentHelper.kt
package com.nytimes.android.external.registerlib

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.ResolveInfo

/**
 * Utility class added to address API 21 issue with calling service with implicit intent
 *
 */
object ServiceIntentHelper {

    /***
     * Android L (lollipop, API 21) introduced a new problem when trying to invoke implicit intent,
     * "java.lang.IllegalArgumentException: Service Intent must be explicit"
     *
     * If you are using an implicit intent, and know only 1 target would answer this intent,
     * This method will help you turn the implicit intent into the explicit form.
     *
     * Inspired from SO answer: http://stackoverflow.com/a/26318757/1446466
     * @param context
     * @param implicitIntent - The original implicit intent
     * @return Explicit Intent created from the implicit original intent
     */
    fun createExplicitFromImplicitIntent(context: Context, implicitIntent: Intent): Intent? {
        val resolveInfo = getResolveInfo(context, implicitIntent) ?: return null

        val component = getComponentFromServiceInfo(resolveInfo)

        return makeExplicitIntent(implicitIntent, component)
    }

    // Retrieve service that can match the given intent
    private fun getResolveInfo(context: Context, implicitIntent: Intent): ResolveInfo? {
        val packageManager = context.packageManager
        val resolveInfoList = packageManager.queryIntentServices(implicitIntent, 0)

        // Make sure only one match was found
        return if (resolveInfoList == null || resolveInfoList.size != 1) {
            null
        } else resolveInfoList[0]
    }

    // Get component info and create ComponentName
    private fun getComponentFromServiceInfo(serviceInfo: ResolveInfo): ComponentName {
        val packageName = serviceInfo.serviceInfo.packageName
        val className = serviceInfo.serviceInfo.name
        return ComponentName(packageName, className)
    }

    private fun makeExplicitIntent(implicitIntent: Intent, component: ComponentName): Intent {
        // Create a new intent. Use the old one for extras and such reuse
        val explicitIntent = Intent(implicitIntent)

        // Set the component to be explicit
        explicitIntent.component = component

        return explicitIntent
    }
}
